### PR TITLE
Prefer "backend" instead of "plugin"

### PR DESCRIPTION
**Pull request by eriknw**
CC @rlratzel, who recently raised the issue of using the term "backend" instead of "plugin".

I think it's reasonable to have a release where we support both "plugin" and "backend" to give backends the most flexibility (we use "plugin" as the attribute we look at and the entry point names).

### CODE REVIEW

**Comment by eriknw** [comment]
I don't understand what mypy is complaining about (as is typical whenever it complains).

File: networkx/utils/backends.py [hunk 1]
```diff
@@ -117,19 +117,35 @@ def _get_plugins(group, *, load_and_call=False):
...
-plugin_info = _get_plugins("networkx.plugin_info", load_and_call=True)
+# Rename "plugin" to "backend", and give backends a release cycle to update.
+backends = _get_backends("networkx.plugins")
+backend_info = _get_backends("networkx.plugin_info", load_and_call=True)
+
+backends.update(_get_backends("networkx.backends"))
+backend_info.update(_get_backends("networkx.backend_info", load_and_call=True))
+
+# Load and cache backends on-demand
+_loaded_backends = {}
```
Comment (jarrodmillman) [inline networkx/utils/backends.py:139]:
```suggestion
_loaded_backends = {}  # type: ignore[var-annotated]
```

Comment (jarrodmillman) [reply networkx/utils/backends.py:139]: That should make the CI happy.

**Comment by jarrodmillman** [comment]
We deleted `tools/pyproject.toml.in` in #6987.

File: networkx/utils/backends.py (hunk 1 above)
Comment (eriknw) [reply networkx/utils/backends.py]: Thanks. Done. Any idea why is this needed here but not for `_registered_algorithms = {}` just a few lines down?

**Comment by eriknw** [comment]
> We deleted tools/pyproject.toml.in in https://github.com/networkx/networkx/pull/6987.

👍. Merged latest main.

**Review by jarrodmillman — APPROVED** [review]

File: networkx/utils/backends.py [hunk 2]
```diff
@@ -117,19 +117,35 @@ def _get_plugins(group, *, load_and_call=False):
...
 
 
-plugins = _get_plugins("networkx.plugins")
-plugin_info = _get_plugins("networkx.plugin_info", load_and_call=True)
+# Rename "plugin" to "backend", and give backends a release cycle to update.
+backends = _get_backends("networkx.plugins")
+backend_info = _get_backends("networkx.plugin_info", load_and_call=True)
+
+backends.update(_get_backends("networkx.backends"))
+backend_info.update(_get_backends("networkx.backend_info", load_and_call=True))
```
Comment (MridulS) [inline networkx/utils/backends.py:136]: I don't think we need to do this, but I guess it's better to do be conservative anyway :)

**Review by MridulS — APPROVED** [review]

**Pull request merged**